File: net/JetBrains.SignatureVerifier/Crypt/SignerInfoVerifier.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using JetBrains.Annotations;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Cms;
using Org.BouncyCastle.Asn1.Ocsp;
using Org.BouncyCastle.Asn1.Oiw;
using Org.BouncyCastle.Asn1.X509;
using Org.BouncyCastle.Cms;
using Org.BouncyCastle.Ocsp;
using Org.BouncyCastle.Pkix;
using Org.BouncyCastle.Security.Certificates;
using Org.BouncyCastle.Tsp;
using Org.BouncyCastle.Utilities.Collections;
using Org.BouncyCastle.Utilities.Date;
using Org.BouncyCastle.X509;
using Org.BouncyCastle.X509.Store;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using AttributeTable = Org.BouncyCastle.Asn1.Cms.AttributeTable;
using CertStatus = Org.BouncyCastle.Asn1.Ocsp.CertStatus;
using CmsSignedData = JetBrains.SignatureVerifier.BouncyCastle.Cms.CmsSignedData;
using SignerInformation = JetBrains.SignatureVerifier.BouncyCastle.Cms.SignerInformation;
using Time = Org.BouncyCastle.Asn1.Cms.Time;
using TimeStampToken = JetBrains.SignatureVerifier.BouncyCastle.Tsp.TimeStampToken;

namespace JetBrains.SignatureVerifier.Crypt
{
  class SignerInfoVerifier
  {
    private readonly SignerInformation _signer;
    private readonly IX509Store _certs;
    private readonly ILogger _logger;
    private TimeStampToken _timeStampToken;
    private List<SignerInformation> _counterSignatures;
    private TimeStampToken TimeStampToken => _timeStampToken ??= getTimeStampToken();
    private List<SignerInformation> CounterSignatures => _counterSignatures ??= getCounterSignatures();
    private CrlProvider _crlProvider;
    private CrlProvider CrlProvider => _crlProvider ??= new CrlProvider(new CrlCacheFileSystem());

    public SignerInfoVerifier([NotNull] SignerInformation signer, [NotNull] IX509Store certs, ILogger logger)
    {
      _signer = signer ?? throw new ArgumentNullException(nameof(signer));
      _certs = certs ?? throw new ArgumentNullException(nameof(certs));
      _logger = logger ?? NullLogger.Instance;
    }

    public async Task<VerifySignatureResult> VerifyAsync(
      [NotNull] SignatureVerificationParams signatureVerificationParams)
    {
      if (signatureVerificationParams == null)
        throw new ArgumentNullException(nameof(signatureVerificationParams));

      var certList = new ArrayList(_certs.GetMatches(_signer.SignerID));

      if (certList.Count < 1)
      {
        _logger.Error(Messages.signer_cert_not_found);

        return new VerifySignatureResult(VerifySignatureStatus.InvalidSignature)
          { Message = Messages.signer_cert_not_found };
      }

      var cert = (X509Certificate)certList[0];

      try
      {
        if (!_signer.Verify(cert))
          return new VerifySignatureResult(VerifySignatureStatus.InvalidSignature);

        var verifyCounterSignResult =
          await verifyCounterSignAsync(signatureVerificationParams);

        if (verifyCounterSignResult.NotValid)
          return verifyCounterSignResult;

        var verifyTimeStampResult = await verifyTimeStampAsync(signatureVerificationParams);

        if (verifyTimeStampResult.NotValid)
          return verifyTimeStampResult;

        var verifyNestedSignsResult = await verifyNestedSignsAsync(signatureVerificationParams);

        if (verifyNestedSignsResult.NotValid)
          return verifyNestedSignsResult;

        if (signatureVerificationParams.BuildChain)
          return await buildCertificateChainAsync(cert, _certs, signatureVerificationParams);

        return VerifySignatureResult.Valid;
      }
      catch (CmsException ex)
      {
        return new VerifySignatureResult(VerifySignatureStatus.InvalidSignature)
          { Message = ex.FlatMessages() };
      }
    }

    private async Task<VerifySignatureResult> verifyNestedSignsAsync(
      SignatureVerificationParams signatureVerificationParams)
    {
      var nestedSignAttrs = _signer.UnsignedAttributes?.GetAll(OIDs.SPC_NESTED_SIGNATURE);

      if (nestedSignAttrs is not null)
      {
        foreach (Attribute nestedSignAttr in nestedSignAttrs)
        {
          if (nestedSignAttr.AttrValues.Count > 0)
          {
            var nestedSignedMessage =
              new SignedMessage(nestedSignAttr.AttrValues[0].ToAsn1Object(), _logger);

            var nestedSignVerifyResult =
              await nestedSignedMessage.VerifySignatureAsync(signatureVerificationParams);

            if (nestedSignVerifyResult.NotValid)
              return nestedSignVerifyResult;
          }
        }
      }

      return VerifySignatureResult.Valid;
    }

    private async Task<VerifySignatureResult> verifyCounterSignAsync(
      SignatureVerificationParams signatureVerificationParams)
    {
      var signerInfoWraps = CounterSignatures.Select(signerInformation =>
        new SignerInfoVerifier(signerInformation, _certs, _logger));

      foreach (var signerInfoWrap in signerInfoWraps)
      {
        var res = await signerInfoWrap.VerifyAsync(signatureVerificationParams);

        if (res.NotValid)
          return res;
      }

      return VerifySignatureResult.Valid;
    }

    private async Task<VerifySignatureResult> verifyTimeStampAsync(
      SignatureVerificationParams signatureVerificationParams)
    {
      var tst = TimeStampToken;

      if (tst == null)
        return VerifySignatureResult.Valid;

      var tstCerts = tst.GetCertificates("Collection");
      var tstCertsList = new ArrayList(tstCerts.GetMatches(tst.SignerID));

      if (tstCertsList.Count < 1)
        return new VerifySignatureResult(VerifySignatureStatus.InvalidTimestamp)
          { Message = Messages.signer_cert_not_found };

      var tstCert = (X509Certificate)tstCertsList[0];

      try
      {
        tst.Validate(tstCert);

        if (signatureVerificationParams.BuildChain)
          try
          {
            var tstCmsSignedData = tst.ToCmsSignedData();
            var certs = tstCmsSignedData.GetCertificates("Collection");
            return await buildCertificateChainAsync(tstCert, certs, signatureVerificationParams);
          }
          catch (PkixCertPathBuilderException ex)
          {
            return VerifySignatureResult.InvalidChain(ex.FlatMessages());
          }
      }
      catch (TspException ex)
      {
        return new VerifySignatureResult(VerifySignatureStatus.InvalidTimestamp)
          { Message = ex.FlatMessages() };
      }
      catch (CertificateExpiredException ex)
      {
        return new VerifySignatureResult(VerifySignatureStatus.InvalidTimestamp)
          { Message = ex.FlatMessages() };
      }

      return VerifySignatureResult.Valid;
    }

    private async Task<VerifySignatureResult> buildCertificateChainAsync(
      X509Certificate primary,
      IX509Store intermediateCertsStore,
      SignatureVerificationParams signatureVerificationParams)
    {
      var builderParams = new CustomPkixBuilderParameters(
        signatureVerificationParams.RootCertificates,
        intermediateCertsStore,
        new X509CertStoreSelector { Certificate = primary },
        getSignValidationTime(signatureVerificationParams));

      var useOCSP = signatureVerificationParams.WithRevocationCheck &&
                    await builderParams.PrepareCrls(CrlProvider);

      try
      {
        var builder = new PkixCertPathBuilder();
        var chain = builder.Build(builderParams);

        if (useOCSP)
        {
          _logger.Trace($"Start OCSP for certificate {primary.FormatId()}");

          var issuerCert = chain.CertPath.Certificates.Cast<X509Certificate>().Last();

          return await new OcspVerifier(signatureVerificationParams.OcspResponseTimeout, _logger)
            .CheckCertificateRevocationStatusAsync(primary, issuerCert);
        }
        else
          return VerifySignatureResult.Valid;
      }
      catch (PkixCertPathBuilderException ex)
      {
        _logger.Error($"Build chain for certificate was failed. {primary.FormatId()} {ex.FlatMessages()}");
        return VerifySignatureResult.InvalidChain(ex.FlatMessages());
      }
    }

    private DateTime? getSignValidationTime(SignatureVerificationParams signatureVerificationParams)
    {
      return signatureVerificationParams.SignValidationTimeMode switch
      {
        SignatureValidationTimeMode.Timestamp => getSigningTime() ?? getTimestamp(),
        SignatureValidationTimeMode.SignValidationTime => signatureVerificationParams.SignatureValidationTime,
        SignatureValidationTimeMode.Current => null,
        _ => throw new ArgumentOutOfRangeException(nameof(signatureVerificationParams.SignValidationTimeMode))
      };
    }

    private List<SignerInformation> getCounterSignatures()
    {
      var res = new List<SignerInformation>();
      addCounterSign(_signer);
      return res;

      void addCounterSign(SignerInformation current)
      {
        foreach (SignerInformation signer in current.GetCounterSignatures().GetSigners())
        {
          res.Add(signer);
          addCounterSign(signer);
        }
      }
    }

    private TimeStampToken getTimeStampToken()
    {
      var timestampAttrValue = getUnsignedAttributeValue(OIDs.MS_COUNTER_SIGN)
                               ?? getUnsignedAttributeValue(OIDs.TIMESTAMP_TOKEN);

      if (timestampAttrValue == null) return null;
      var contentInfo = ContentInfo.GetInstance(timestampAttrValue);
      var cmsSignedData = new CmsSignedData(contentInfo);
      return new TimeStampToken(cmsSignedData);
    }

    private DateTime? getTimestamp() => TimeStampToken?.TimeStampInfo.GenTime ?? getTimeStampFromCounterSign();

    private DateTime? getTimeStampFromCounterSign()
    {
      var res = CounterSignatures.Select(signer =>
      {
        var signingTimeAttribute = signer.SignedAttributes?[OIDs.SIGNING_TIME];

        if (signingTimeAttribute is not null && signingTimeAttribute.AttrValues.Count > 0)
        {
          var attrValue = signingTimeAttribute.AttrValues[0];
          var time = Time.GetInstance(attrValue);
          return time.Date;
        }

        return (DateTime?)null;
      }).FirstOrDefault(f => f.HasValue);

      return res;
    }

    private DateTime? getSigningTime()
    {
      var signingTime = getSignedAttributeValue(CmsAttributes.SigningTime);
      return signingTime == null ? null : Time.GetInstance(signingTime).Date;
    }

    private Asn1Encodable getSignedAttributeValue(DerObjectIdentifier oid) =>
      _signer.SignedAttributes?.GetFirstAttributeValue(oid);

    private Asn1Encodable getUnsignedAttributeValue(DerObjectIdentifier oid) =>
      _signer.UnsignedAttributes?.GetFirstAttributeValue(oid);
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/crypt/SignerInfoVerifier.kt
package com.jetbrains.signatureverifier.crypt

import com.jetbrains.signatureverifier.ILogger
import com.jetbrains.signatureverifier.Messages
import com.jetbrains.signatureverifier.NullLogger
import com.jetbrains.signatureverifier.crypt.BcExt.FormatId
import com.jetbrains.signatureverifier.crypt.BcExt.GetFirstAttributeValue
import com.jetbrains.signatureverifier.crypt.BcExt.ToJavaX509Certificate
import com.jetbrains.signatureverifier.crypt.BcExt.ToX509CertificateHolder
import com.jetbrains.signatureverifier.crypt.Utils.ConvertToLocalDateTime
import com.jetbrains.signatureverifier.crypt.Utils.FlatMessages
import com.jetbrains.signatureverifier.crypt.Utils.ToString
import com.jetbrains.signatureverifier.crypt.Utils.isExceptionIgnored
import org.bouncycastle.asn1.ASN1Encodable
import org.bouncycastle.asn1.ASN1EncodableVector
import org.bouncycastle.asn1.ASN1ObjectIdentifier
import org.bouncycastle.asn1.cms.Attribute
import org.bouncycastle.asn1.cms.CMSAttributes
import org.bouncycastle.asn1.cms.ContentInfo
import org.bouncycastle.asn1.cms.Time
import org.bouncycastle.cert.X509CertificateHolder
import org.bouncycastle.cms.CMSException
import org.bouncycastle.cms.CMSSignedData
import org.bouncycastle.cms.SignerInformation
import org.bouncycastle.cms.jcajce.JcaSignerInfoVerifierBuilder
import org.bouncycastle.operator.jcajce.JcaDigestCalculatorProviderBuilder
import org.bouncycastle.tsp.TSPException
import org.bouncycastle.tsp.TimeStampToken
import org.bouncycastle.util.Selector
import org.bouncycastle.util.Store
import org.jetbrains.annotations.NotNull
import java.security.cert.*
import java.util.*

open class SignerInfoVerifier(
  @NotNull signer: SignerInformation,
  @NotNull certs: Store<X509CertificateHolder>,
  @NotNull crlProvider: CrlProvider,
  logger: ILogger?
) {
  private val _signer: SignerInformation = signer
  private val _certs: Store<X509CertificateHolder> = certs
  private val _crlProvider: CrlProvider = crlProvider
  private val _logger: ILogger
  private val TimeStampToken by lazy { timeStampToken() }
  private val CounterSignatures by lazy { counterSignatures() }

  init {
    _logger = logger ?: NullLogger.Instance
  }

  suspend fun VerifyAsync(@NotNull params: SignatureVerificationParams): VerifySignatureResult {
    val certList = ArrayList(_certs.getMatches(_signer.sid as Selector<X509CertificateHolder>))
    if (certList.isEmpty()) {
      _logger.Error(Messages.signer_cert_not_found)
      return VerifySignatureResult(VerifySignatureStatus.InvalidSignature, Messages.signer_cert_not_found, null)
    }
    val cert = certList[0] as X509CertificateHolder
    try {
      val verifier = JcaSignerInfoVerifierBuilder(JcaDigestCalculatorProviderBuilder().build()).build(cert)

      val verify = _signer.verify(verifier)
      if (!verify) {
        Utils.printCertificateError(cert,params.testedFileName)
        return VerifySignatureResult(VerifySignatureStatus.InvalidSignature, certificate = cert)
      }

      if (params.BuildChain)
        applySignValidationTime(params)

      val verifyCounterSignResult = verifyCounterSignAsync(params)
      if (verifyCounterSignResult.NotValid)
        return verifyCounterSignResult

      val verifyTimeStampResult = verifyTimeStampAsync(params)
      if (verifyTimeStampResult.NotValid)
        return verifyTimeStampResult

      val verifyNestedSignsResult = verifyNestedSignsAsync(params)
      if (verifyNestedSignsResult.NotValid)
        return verifyNestedSignsResult

      if (params.BuildChain)
        return buildCertificateChainAsync(cert, _certs, params)

      return VerifySignatureResult.Valid
    } catch (ex: CMSException) {
      Utils.printCertificateError(cert,params.testedFileName)
      return if (isExceptionIgnored(ex)) VerifySignatureResult(
        params.expectedResult,
        certificate = null
      ) else VerifySignatureResult(
        VerifySignatureStatus.InvalidSignature,
        ex.FlatMessages(),
        null
      )
    } catch (ex: CertificateExpiredException) {
      return VerifySignatureResult(VerifySignatureStatus.InvalidSignature, ex.FlatMessages(), null)
    }
  }

  private fun applySignValidationTime(signatureVerificationParams: SignatureVerificationParams) {
    if (signatureVerificationParams.SignValidationTimeMode != SignatureValidationTimeMode.Timestamp || signatureVerificationParams.SignatureValidationTime != null)
      return

    val signValidationTime = getSigningTime() ?: getTimestamp()
    if (signValidationTime != null)
      signatureVerificationParams.SetSignValidationTime(signValidationTime.ConvertToLocalDateTime())
    else
      _logger.Warning("Unknown sign validation time")
  }

  private suspend fun verifyNestedSignsAsync(signatureVerificationParams: SignatureVerificationParams): VerifySignatureResult {
    val verifyNestedSignsResult = verifyNestedSignsAsync(OIDs.SPC_NESTED_SIGNATURE, signatureVerificationParams)
    if (verifyNestedSignsResult.NotValid)
      return verifyNestedSignsResult

    val verifyMsCounterSignsResult = verifyNestedSignsAsync(OIDs.MS_COUNTER_SIGN, signatureVerificationParams)
    if (verifyMsCounterSignsResult.NotValid)
      return verifyMsCounterSignsResult

    return VerifySignatureResult.Valid
  }

  private suspend fun verifyNestedSignsAsync(
    attrOid: ASN1ObjectIdentifier,
    signatureVerificationParams: SignatureVerificationParams
  ): VerifySignatureResult {
    val nestedSignAttrs = _signer.unsignedAttributes?.getAll(attrOid) ?: return VerifySignatureResult.Valid

    for (_nestedSignAttr in nestedSignAttrs) {
      val nestedSignAttr = _nestedSignAttr as Attribute
      for (attrValue in nestedSignAttr.attrValues) {
        val nestedSignedMessage = SignedMessage(attrValue.toASN1Primitive())
        val nestedSignVerifyResult = SignedMessageVerifier(_crlProvider, _logger).VerifySignatureAsync(
          nestedSignedMessage,
          signatureVerificationParams
        )
        if (nestedSignVerifyResult.NotValid)
          return nestedSignVerifyResult
      }
    }
    return VerifySignatureResult.Valid
  }

  private suspend fun verifyCounterSignAsync(signatureVerificationParams: SignatureVerificationParams): VerifySignatureResult {
    val signerInfoWraps = CounterSignatures.map { signerInformation ->
      SignerInfoVerifier(
        signerInformation,
        _certs,
        _crlProvider,
        _logger
      )
    }
    for (signerInfoWrap in signerInfoWraps) {
      val res = signerInfoWrap.VerifyAsync(signatureVerificationParams)
      if (res.NotValid)
        return res
    }
    return VerifySignatureResult.Valid
  }

  private suspend fun verifyTimeStampAsync(params: SignatureVerificationParams): VerifySignatureResult {
    val tst = TimeStampToken ?: return VerifySignatureResult.Valid
    val tstCerts = tst.certificates
    val tstCertsList = ArrayList(tstCerts.getMatches(tst.sid as Selector<X509CertificateHolder>))
    if (tstCertsList.isEmpty())
      return VerifySignatureResult(VerifySignatureStatus.InvalidTimestamp, Messages.signer_cert_not_found, null)

    val tstCert = tstCertsList[0] as X509CertificateHolder
    try {
      val verifier = JcaSignerInfoVerifierBuilder(JcaDigestCalculatorProviderBuilder().build()).build(tstCert)
      tst.validate(verifier)
      if (params.BuildChain)
        return try {
          val tstCmsSignedData = tst.toCMSSignedData()
          val certs = tstCmsSignedData.certificates
          buildCertificateChainAsync(tstCert, certs, params)
        } catch (ex: CertPathBuilderException) {
          VerifySignatureResult.InvalidChain(ex.FlatMessages())
        }
    } catch (ex: TSPException) {
      Utils.printCertificateError(tstCert,params.testedFileName)
      return if (isExceptionIgnored(ex)) VerifySignatureResult.Valid else
        VerifySignatureResult(
          VerifySignatureStatus.InvalidTimestamp,
          ex.FlatMessages(),
          null
        )
    } catch (ex: CertificateExpiredException) {
      return VerifySignatureResult(VerifySignatureStatus.InvalidTimestamp, ex.FlatMessages(), null)
    }
    return VerifySignatureResult.Valid
  }

  private suspend fun buildCertificateChainAsync(
    primary: X509CertificateHolder,
    intermediateCertsStore: Store<X509CertificateHolder>,
    signatureVerificationParams: SignatureVerificationParams
  ): VerifySignatureResult {
    _logger.Trace("Signature validation time: ${signatureVerificationParams.SignatureValidationTime?.ToString("dd.MM.uuuu HH:mm:ss") ?: "<null>"}")

    val builderParams = CustomPkixBuilderParameters(
      signatureVerificationParams.RootCertificates!!.toHashSet(),
      intermediateCertsStore,
      X509CertSelector().also { it.certificate = primary.ToJavaX509Certificate() },
      signatureVerificationParams.SignatureValidationTime
    )

    val useOCSP = signatureVerificationParams.WithRevocationCheck &&
      builderParams.PrepareCrls(_crlProvider)

    try {
      val builder = CertPathBuilder.getInstance(CertPathBuilder.getDefaultType())
      val chain = builder.build(builderParams) as PKIXCertPathBuilderResult

      if (useOCSP) {
        _logger.Trace("Start OCSP for certificate ${primary.FormatId()}")
        val issuerCert = getIssuerCert(chain, primary)
        return OcspVerifier(signatureVerificationParams.OcspResponseTimeout, _logger)
          .CheckCertificateRevocationStatusAsync(primary, issuerCert!!)
      }

      return VerifySignatureResult.Valid
    } catch (ex: CertPathBuilderException) {
      _logger.Error("Build chain for certificate was failed. ${primary.FormatId()} ${ex.FlatMessages()}")
      return VerifySignatureResult.InvalidChain(ex.FlatMessages())
    }
  }

  private fun getIssuerCert(chain: PKIXCertPathBuilderResult, cert: X509CertificateHolder): X509CertificateHolder? {
    return chain.certPath.certificates.map { it.ToX509CertificateHolder() }
      .lastOrNull { it.subject.equals(cert.issuer) }
      ?: chain.trustAnchor.trustedCert?.ToX509CertificateHolder()
  }

  private fun counterSignatures(): Collection<SignerInformation> {
    val res = mutableListOf<SignerInformation>()

    fun addCounterSign(current: SignerInformation) {
      for (signer in current.getCounterSignatures().signers) {
        res.add(signer)
        addCounterSign(signer)
      }
    }
    addCounterSign(_signer)
    return res
  }

  private fun timeStampToken(): TimeStampToken? {
    val timestampAttrValue =
      getUnsignedAttributeValue(OIDs.MS_COUNTER_SIGN) ?: getUnsignedAttributeValue(OIDs.TIMESTAMP_TOKEN) ?: return null
    val contentInfo = ContentInfo.getInstance(timestampAttrValue)
    val cmsSignedData = CMSSignedData(contentInfo)
    return TimeStampToken(cmsSignedData)
  }

  private fun getTimestamp(): Date? = TimeStampToken?.timeStampInfo?.genTime ?: getTimeStampFromCounterSign()

  private fun getTimeStampFromCounterSign(): Date? {
    val items: List<Date?> = CounterSignatures.map { signer ->
      val signingTimeAttribute = signer.signedAttributes?.get(OIDs.SIGNING_TIME)
      if (signingTimeAttribute != null && signingTimeAttribute.attrValues.count() > 0) {
        val attrValue = signingTimeAttribute.attrValues.getObjectAt(0)
        val time = Time.getInstance(attrValue)
        return time.date
      }
      return null
    }
    return items.firstOrNull { f -> f != null }
  }

  private fun getSigningTime(): Date? {
    val signingTime = getSignedAttributeValue(CMSAttributes.signingTime)
    return if (signingTime == null) null else Time.getInstance(signingTime).date
  }

  private fun getSignedAttributeValue(oid: ASN1ObjectIdentifier): ASN1Encodable? =
    _signer.signedAttributes?.GetFirstAttributeValue(oid)

  private fun getUnsignedAttributeValue(oid: ASN1ObjectIdentifier): ASN1Encodable? =
    _signer.unsignedAttributes?.GetFirstAttributeValue(oid)

  private operator fun ASN1EncodableVector.iterator(): ASN1EncodableVectorIterator {
    return ASN1EncodableVectorIterator(this)
  }
}


